### PR TITLE
Create a primitive sampling showcase example

### PR DESCRIPTION
**Pull request by IQuick143**
# Objective

- Show + Visually Test that 3D primitive sampling works
- Make an example that looks nice.

## Solution

- Added a `sampling_primitives` examples which shows all the 3D primitives being sampled, with a firefly aesthetic.

image

## Testing

- `cargo run --example sampling_primitives`
- Haven't tested WASM.

### Added
- Added a new example, `sampling_primitives`, to showcase all the 3D sampleable primitives.

## Additional notes:

This example borrowed a bunch of code from the other sampling example, by @mweatherley.

In future updates this example should be updated with new 3D primitives as they become sampleable.

### CODE REVIEW

**Comment by IQuick143** [comment]
Please give suggestions on code quality, visual and performance improvements.

File: examples/math/sampling_primitives.rs [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+        let n_shapes = shapes.len();
+
+        let translations =
+            (0..n_shapes).map(|i| (i as f32 - n_shapes as f32 / 2.0) * DISTANCE_BETWEEN_SHAPES);
+
+        SampledShapes(shapes.into_iter().zip(translations).collect())
+    }
+}
+
+/// Enum wrapping the individual
```
Comment (alice-i-cecile) [inline examples/math/sampling_primitives.rs:145]:
```suggestion
/// Enum listing the shapes that can be sampled
```

File: examples/math/sampling_primitives.rs [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+#[derive(Resource)]
+struct PointCounter(usize);
+
+/// Resource storing the shapes being sampled and their translations.
+#[derive(Resource)]
+struct SampledShapes(Vec<(Shape, Vec3)>);
+
+impl SampledShapes {
+    fn new() -> Self {
+        let shapes = vec![
```
Comment (alice-i-cecile) [inline examples/math/sampling_primitives.rs:127]: This list should be a method on `Shape`.

File: examples/math/sampling_primitives.rs [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+                },
+                transform: Transform::from_translation(sample),
+                ..default()
+            },
+            SamplePoint,
+        ));
+    }
+}
+
+fn despawn_points(
```
Comment (alice-i-cecile) [inline examples/math/sampling_primitives.rs:578]: Visually, I think that fading these points out before despawn would work great.

Comment (IQuick143) [reply examples/math/sampling_primitives.rs:578]: Agreed, fading in would also be really neat imo.
I didn't know the best way to do so and I didn't want to tack-on another hack so I left it out.
If anyone has an idea how to do it neatly I'll add it in.

Comment (alice-i-cecile) [reply examples/math/sampling_primitives.rs:578]: Mutate the light's intensity over time using a system that checks the delta-time. And then when the intensity reaches 0 despawn them.

Comment (IQuick143) [reply examples/math/sampling_primitives.rs:585]: The issue is that the lights are not actually lights on the points, they're a combination of larger lights (which aggregate the "light" from all the points) and emissive colours on the Material (which is shared among all the points for performance and simplicity).

What could work is if the points were scaled to "pop in" or sizzle out, they'd get a special component for doing so.

I'll take a look on how difficult that would be, but it sounds reasonable.

**Review by alice-i-cecile — APPROVED** [review]

**Review by Jondolf — COMMENTED** [review]
Looks good! Might be almost too fancy for just demonstrating sampling, but it does showcase it well and is visually appealing :) A screenshot or video of this would be a great fit for the 0.14 release notes too!

Left just a few non-blocking nits. Also the keyboard inputs for zooming are sort of flipped and/or partially broken like we discussed on Discord, and for example pressing the non-numpad version of the minus/hyphen doesn't work with my Nordic (Finnish) keyboard layout, but I probably wouldn't block over it

File: examples/math/sampling_primitives.rs [hunk 4]
```diff
@@ -0,0 +1,698 @@
...
+        Vec3::new(0.0, -BIG_3D * 0.67, -BIG_3D * 1.17),
+        Vec3::new(0.0, BIG_3D, 0.0),
+    ],
+};
+
+// Components, Resources
+
+/// Resource for the random sampling mode, telling whether to sample the interior or the boundary.
+#[derive(Resource)]
+enum Mode {
```
[the remark this answers is not in the document]
Comment (IQuick143) [reply examples/math/sampling_primitives.rs]: Good call.

**Review by Jondolf — APPROVED** [review]
LGTM! The input issue seems to just be that the physical keys (`KeyCode`) don't map to the correct logical keys with my keyboard layout, but I don't think we have a good clean approach to do it properly, and the other examples also just use key codes.

**Pull request merged**